# Livro para consulta:
- https://jakevdp.github.io/PythonDataScienceHandbook/03.08-aggregation-and-grouping.html
- https://jakevdp.github.io/PythonDataScienceHandbook/03.09-pivot-tables.html
    

# 1. Importando bibliotecas <a name="import"></a>

<div style="text-align: right"
     
[Voltar ao índice](#Contents)

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

%matplotlib inline

# 2. Carregando o dataframe SINASC <a name="read"></a>
<div style="text-align: right"
     
[Voltar ao índice](#Contents)

In [2]:
sinasc_raw = pd.read_csv('SINASC_RO_2019.csv')
sinasc_raw.head()

ORIGEM   CODESTAB  CODMUNNASC  LOCNASC  IDADEMAE  ESTCIVMAE  \
0       1  2679477.0      110001        1        19        5.0   
1       1  2679477.0      110001        1        29        2.0   
2       1  2679477.0      110001        1        37        9.0   
3       1  2516500.0      110001        1        30        5.0   
4       1  2516500.0      110001        1        30        2.0   

            ESCMAE  CODOCUPMAE  QTDFILVIVO  QTDFILMORT  ...  KOTELCHUCK  \
0      8 a 11 anos         NaN         0.0         0.0  ...           5   
1      8 a 11 anos    999992.0         1.0         0.0  ...           5   
2      8 a 11 anos    513205.0         2.0         0.0  ...           5   
3  12 anos ou mais    231205.0         0.0         0.0  ...           4   
4      8 a 11 anos    999992.0         1.0         0.0  ...           5   

  CONTADOR munResStatus munResTipo               munResNome  munResUf  \
0        1        ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia   
1        2        ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia   
2        3        ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia   
3        4        ATIVO      MUNIC  Alto Alegre dos Parecis  Rondônia   
4        5        ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia   

   munResLat munResLon  munResAlt  munResArea  
0  -11.93554 -61.99982      338.0    7067.025  
1  -11.93554 -61.99982      338.0    7067.025  
2  -11.93554 -61.99982      338.0    7067.025  
3  -12.13178 -61.85308      397.0    3958.273  
4  -11.93554 -61.99982      338.0    7067.025  

[5 rows x 69 columns]

In [3]:
sinasc_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27028 entries, 0 to 27027
Data columns (total 69 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   ORIGEM        27028 non-null  int64  
 1   CODESTAB      26913 non-null  float64
 2   CODMUNNASC    27028 non-null  int64  
 3   LOCNASC       27028 non-null  int64  
 4   IDADEMAE      27028 non-null  int64  
 5   ESTCIVMAE     26711 non-null  float64
 6   ESCMAE        26716 non-null  object 
 7   CODOCUPMAE    24121 non-null  float64
 8   QTDFILVIVO    25455 non-null  float64
 9   QTDFILMORT    24930 non-null  float64
 10  CODMUNRES     27028 non-null  int64  
 11  GESTACAO      25796 non-null  object 
 12  GRAVIDEZ      26949 non-null  object 
 13  PARTO         26979 non-null  object 
 14  CONSULTAS     27028 non-null  int64  
 15  DTNASC        27028 non-null  object 
 16  HORANASC      27007 non-null  float64
 17  SEXO          27024 non-null  object 
 18  APGAR1        26932 non-nu

# Tarefa 1

### 1. Idade media das mães e dos pais por município (coluna munResNome)


In [4]:
sinasc_1 = sinasc_raw[['munResNome', 'IDADEMAE', 'IDADEPAI']]
sinasc_1.head()

munResNome  IDADEMAE  IDADEPAI
0    Alta Floresta D'Oeste        19      26.0
1    Alta Floresta D'Oeste        29      24.0
2    Alta Floresta D'Oeste        37      32.0
3  Alto Alegre dos Parecis        30      24.0
4    Alta Floresta D'Oeste        30      27.0

In [5]:
resultado_1 = sinasc_1.groupby('munResNome')['IDADEMAE', 'IDADEPAI'].mean()
resultado_1

C:\Users\lucca\AppData\Local\Temp\ipykernel_22496\2437675915.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  resultado_1 = sinasc_1.groupby('munResNome')['IDADEMAE', 'IDADEPAI'].mean()


IDADEMAE   IDADEPAI
munResNome                                     
Alta Floresta D'Oeste      25.991826  29.548701
Alto Alegre dos Parecis    24.844156  29.179104
Alto Paraíso               24.960177  28.833333
Alvorada D'Oeste           25.770732  30.757282
Ariquemes                  25.607866  32.466667
Buritis                    25.559177  30.812500
Cabixi                     26.037500  34.333333
Cacaulândia                25.546667  36.333333
Cacoal                     26.890102  30.931854
Campo Novo de Rondônia     24.825301  30.500000
Candeias do Jamari         25.243323  29.894737
Castanheiras               27.281250  30.392857
Cerejeiras                 27.240506  31.652778
Chupinguaia                25.333333  29.629630
Colorado do Oeste          27.602804  34.354839
Corumbiara                 24.716535  32.470588
Costa Marques              24.412017  30.613636
Cujubim                    24.482927  31.400000
Espigão D'Oeste            26.093333  31.198113
Governador Jorge Teixeira  24.401786  29.222222
Guajará-Mirim              25.462228  30.470588
Itapuã do Oeste            26.064220  42.500000
Jaru                       25.982974  32.197183
Ji-Paraná                  26.361137  31.132239
Machadinho D'Oeste         25.480000  31.326923
Ministro Andreazza         27.904762  33.849315
Mirante da Serra           26.304878  32.154762
Monte Negro                25.648241  34.678571
Município ignorado - RO    24.000000  22.000000
Nova Brasilândia D'Oeste   26.222222  30.520833
Nova Mamoré                25.326389  31.692308
Nova União                 26.894231  31.371429
Novo Horizonte do Oeste    26.938776  31.588235
Ouro Preto do Oeste        26.375594  31.233244
Parecis                    26.454545  32.473684
Pimenta Bueno              25.199675  29.745174
Pimenteiras do Oeste       25.075000  35.600000
Porto Velho                26.312670  32.352679
Presidente Médici          25.796053  30.150754
Primavera de Rondônia      27.023256  31.888889
Rio Crespo                 26.660000  28.000000
Rolim de Moura             26.600863  30.360479
Santa Luzia D'Oeste        26.121212  30.703297
Seringueiras               24.910714  30.500000
São Felipe D'Oeste         26.481481  31.395833
São Francisco do Guaporé   25.660066  31.638889
São Miguel do Guaporé      25.926346  32.666667
Teixeirópolis              26.031250  31.500000
Theobroma                  25.536082  30.083333
Urupá                      26.688235  31.008929
Vale do Anari              26.096774  34.000000
Vale do Paraíso            25.679012  29.976744
Vilhena                    26.433962  34.152941

In [32]:
resultado.plot(figsize=[18,6])
plt.ylabel('idade média dos pais');

NameError: name 'resultado' is not defined

Ao analisar o gráfico ficou comprovado que em apenas uma cidade a idade média dos pais era menor que a idade média das mães. Isso ocorreu na célula onde o lançamento do dado foi ignorado no ato do lançamento.

### 2. Peso médio dos bebes por sexo que nasceram no dia do seu aniversário por faixas de escolaridade mae
Ex: Você, aluna(o), nasceu no dia 10/01, então você precisa filtrar o conjunto de dados nessa data e calcular o peso médio dos bebês de cada sexo por faixa de escolaridade da mãe.

In [33]:
sinasc_2 = sinasc_raw[['ESCMAE', 'SEXO', 'PESO', 'DTNASC']]
sinasc_2.head()

ESCMAE       SEXO  PESO      DTNASC
0      8 a 11 anos  Masculino  3685  2019-02-19
1      8 a 11 anos   Feminino  3055  2019-02-21
2      8 a 11 anos   Feminino  3460  2019-02-25
3  12 anos ou mais   Feminino  2982  2019-03-20
4      8 a 11 anos   Feminino  3310  2019-03-23

In [34]:
resultado_2 = sinasc_2[sinasc_2['DTNASC'] == '2019-06-03'].groupby(['ESCMAE', 'SEXO'])['PESO'].mean().unstack()
resultado_2

SEXO                Feminino    Masculino
ESCMAE                                   
1 a 3 anos               NaN  3750.000000
12 anos ou mais  2957.500000  3366.785714
4 a 7 anos       3081.666667  3094.000000
8 a 11 anos      3204.391304  3209.652174

### 3. Qual o municipio que nasceu menos bebe em 2019?
    - qual a idade media, maxima, minima das maes nesse municipio?
    - qual a idade media, maxima, minima dos pais nesse municipio?

In [35]:
sinasc_3 = sinasc_raw[['munResNome', 'IDADEMAE', 'IDADEPAI']]
sinasc_3.head()

munResNome  IDADEMAE  IDADEPAI
0    Alta Floresta D'Oeste        19      26.0
1    Alta Floresta D'Oeste        29      24.0
2    Alta Floresta D'Oeste        37      32.0
3  Alto Alegre dos Parecis        30      24.0
4    Alta Floresta D'Oeste        30      27.0

In [36]:
sinasc_3['munResNome'].value_counts()

Porto Velho                  8437
Ji-Paraná                    2182
Ariquemes                    1729
Vilhena                      1590
Cacoal                       1374
Rolim de Moura                927
Jaru                          881
Guajará-Mirim                 781
Ouro Preto do Oeste           631
Pimenta Bueno                 616
Buritis                       583
Machadinho D'Oeste            500
Espigão D'Oeste               450
Nova Mamoré                   432
Alta Floresta D'Oeste         367
São Miguel do Guaporé         353
Candeias do Jamari            337
Presidente Médici             304
São Francisco do Guaporé      303
Nova Brasilândia D'Oeste      261
Cerejeiras                    237
Costa Marques                 233
Alto Paraíso                  226
Colorado do Oeste             214
Alvorada D'Oeste              205
Cujubim                       205
Monte Negro                   199
Urupá                         170
Seringueiras                  168
Campo Novo de 

O município com maior número de nascimento (8437) foi Porto Velho.

In [37]:
resultado_3 =  sinasc_3[sinasc_3['munResNome'] == 'Porto Velho'] .groupby('munResNome').aggregate(['min', np.median, max])

In [38]:
resultado_3

IDADEMAE            IDADEPAI             
                 min median max      min median   max
munResNome                                           
Porto Velho       12   26.0  47     16.0   32.0  65.0

### 4. Qual o municipio que nasceu mais bebe no mês de março?
    - qual a quantidade de filhos vivos media, maxima, minima nesse municipio?
    - qual a idade media, maxima, minima dos pais nesse municipio?



In [52]:
sinasc_4 = sinasc_raw[['munResNome', 'DTNASC', 'IDADEPAI', 'QTDFILVIVO']]
sinasc_4.head()

munResNome      DTNASC  IDADEPAI  QTDFILVIVO
0    Alta Floresta D'Oeste  2019-02-19      26.0         0.0
1    Alta Floresta D'Oeste  2019-02-21      24.0         1.0
2    Alta Floresta D'Oeste  2019-02-25      32.0         2.0
3  Alto Alegre dos Parecis  2019-03-20      24.0         0.0
4    Alta Floresta D'Oeste  2019-03-23      27.0         1.0

In [102]:
março_19 = sinasc_4.loc[(sinasc_4['DTNASC'] >= '2019-03-01')
                     & (sinasc_4['DTNASC'] < '2019-03-31')]
março_19[('munResNome')].value_counts()

Porto Velho                  727
Ji-Paraná                    180
Vilhena                      146
Ariquemes                    139
Cacoal                       132
Jaru                          78
Guajará-Mirim                 73
Rolim de Moura                69
Ouro Preto do Oeste           67
Pimenta Bueno                 61
Buritis                       57
Nova Mamoré                   52
Espigão D'Oeste               43
Machadinho D'Oeste            41
São Miguel do Guaporé         40
Candeias do Jamari            33
Alta Floresta D'Oeste         31
São Francisco do Guaporé      30
Presidente Médici             29
Costa Marques                 23
Alto Paraíso                  22
Alvorada D'Oeste              22
Monte Negro                   20
Nova Brasilândia D'Oeste      19
Colorado do Oeste             19
Seringueiras                  19
Chupinguaia                   18
Cujubim                       17
Cerejeiras                    16
Urupá                         14
Vale do Pa

In [103]:
resultado_4 =  março_19[março_19['munResNome'] == 'Porto Velho'] .groupby('munResNome').aggregate(['min', np.median, max])

C:\Users\lucca\AppData\Local\Temp\ipykernel_23620\2996705496.py:1: FutureWarning: ['DTNASC'] did not aggregate successfully. If any error is raised this will raise in a future version of pandas. Drop these columns/ops to avoid this warning.
  resultado_4 =  março_19[março_19['munResNome'] == 'Porto Velho'] .groupby('munResNome').aggregate(['min', np.median, max])


In [104]:
resultado_4

IDADEPAI              QTDFILVIVO            
                 min median   max        min median  max
munResNome                                              
Porto Velho     19.0   34.0  62.0        0.0    1.0  9.0

O resultado da análise mostra que a idade do pai pode variar de 19 a 62 anos e a quantidade de filhos vivos varia de nenhum filho até 9 filhos.

### Analise as respostas encontradas, tire algum insight delas, conte pra gente algo encontrado nos dados. Algo que você julgue relevante e novo pra você.

Exemplo:
- Ah, descobri que a idade mediana das mulheres que deram a luz no ano de 2019 dos municipios x é maior que y.

In [15]:
sinasc_raw.dtypes

ORIGEM          int64
CODESTAB      float64
CODMUNNASC      int64
LOCNASC         int64
IDADEMAE        int64
               ...   
munResUf       object
munResLat     float64
munResLon     float64
munResAlt     float64
munResArea    float64
Length: 69, dtype: object

In [16]:
sinasc_raw.select_dtypes(exclude=['int64','float64'])

ESCMAE         GESTACAO GRAVIDEZ    PARTO      DTNASC  \
0          8 a 11 anos  37 a 41 semanas    Única  Cesáreo  2019-02-19   
1          8 a 11 anos  37 a 41 semanas    Única  Cesáreo  2019-02-21   
2          8 a 11 anos  37 a 41 semanas    Única  Cesáreo  2019-02-25   
3      12 anos ou mais  37 a 41 semanas    Única  Cesáreo  2019-03-20   
4          8 a 11 anos  37 a 41 semanas    Única  Cesáreo  2019-03-23   
...                ...              ...      ...      ...         ...   
27023  12 anos ou mais  32 a 36 semanas    Única  Cesáreo  2019-12-13   
27024      8 a 11 anos  37 a 41 semanas    Única  Vaginal  2019-10-04   
27025      8 a 11 anos  37 a 41 semanas    Única  Cesáreo  2019-08-02   
27026      8 a 11 anos  32 a 36 semanas    Única  Vaginal  2019-12-23   
27027      8 a 11 anos  37 a 41 semanas    Única  Cesáreo  2019-12-27   

            SEXO RACACOR  DTCADASTRO CODANOMAL VERSAOSIST   DTRECEBIM  \
0      Masculino  Branca  2019-03-08       NaN     3.2.01  2019-03-08   
1       Feminino  Branca  2019-03-08       NaN     3.2.01  2019-03-08   
2       Feminino  Branca  2019-03-08       NaN     3.2.01  2019-03-08   
3       Feminino   Parda  2019-04-03       NaN     3.2.01  2019-04-09   
4       Feminino   Parda  2019-04-03       NaN     3.2.01  2019-04-09   
...          ...     ...         ...       ...        ...         ...   
27023  Masculino  Branca  2020-09-28       NaN     3.2.01  2020-10-13   
27024  Masculino  Branca  2019-10-16       NaN     3.2.01  2019-10-21   
27025  Masculino   Parda  2019-08-07       NaN     3.2.00  2019-08-07   
27026  Masculino   Parda  2020-01-07       NaN     3.2.00  2020-01-07   
27027  Masculino   Parda  2020-10-15       NaN     3.2.01  2020-10-16   

      munResStatus munResTipo               munResNome  munResUf  
0            ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia  
1            ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia  
2            ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia  
3            ATIVO      MUNIC  Alto Alegre dos Parecis  Rondônia  
4            ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia  
...            ...        ...                      ...       ...  
27023        ATIVO      MUNIC                  Vilhena  Rondônia  
27024        ATIVO      MUNIC              Chupinguaia  Rondônia  
27025        ATIVO      MUNIC                  Vilhena  Rondônia  
27026        ATIVO      MUNIC                  Vilhena  Rondônia  
27027        ATIVO      MUNIC  Novo Horizonte do Oeste  Rondônia  

[27028 rows x 15 columns]

In [17]:
sinasc_raw.columns

Index(['ORIGEM', 'CODESTAB', 'CODMUNNASC', 'LOCNASC', 'IDADEMAE', 'ESTCIVMAE',
       'ESCMAE', 'CODOCUPMAE', 'QTDFILVIVO', 'QTDFILMORT', 'CODMUNRES',
       'GESTACAO', 'GRAVIDEZ', 'PARTO', 'CONSULTAS', 'DTNASC', 'HORANASC',
       'SEXO', 'APGAR1', 'APGAR5', 'RACACOR', 'PESO', 'IDANOMAL', 'DTCADASTRO',
       'CODANOMAL', 'NUMEROLOTE', 'VERSAOSIST', 'DTRECEBIM', 'DIFDATA',
       'DTRECORIGA', 'NATURALMAE', 'CODMUNNATU', 'CODUFNATU', 'ESCMAE2010',
       'SERIESCMAE', 'DTNASCMAE', 'RACACORMAE', 'QTDGESTANT', 'QTDPARTNOR',
       'QTDPARTCES', 'IDADEPAI', 'DTULTMENST', 'SEMAGESTAC', 'TPMETESTIM',
       'CONSPRENAT', 'MESPRENAT', 'TPAPRESENT', 'STTRABPART', 'STCESPARTO',
       'TPNASCASSI', 'TPFUNCRESP', 'TPDOCRESP', 'DTDECLARAC', 'ESCMAEAGR1',
       'STDNEPIDEM', 'STDNNOVA', 'CODPAISRES', 'TPROBSON', 'PARIDADE',
       'KOTELCHUCK', 'CONTADOR', 'munResStatus', 'munResTipo', 'munResNome',
       'munResUf', 'munResLat', 'munResLon', 'munResAlt', 'munResArea'],
      dtype='object')